lib/SIL/TypeLowering.cpp

```diff
@@ -1424,7 +1425,24 @@ static CanTupleType computeLoweredTupleType(TypeConverter &tc,
                                             TypeExpansionContext context,
                                             AbstractionPattern origType,
                                             CanTupleType substType) {
-  assert(origType.matchesTuple(substType));
+
+  // Don't trigger this assertion for () when it's being returned as an opaque
+  // return for `some Equatable`
+  bool isOpaqueEquatableVoid = false;
+  if (!origType.isTypeParameter()) {
+    auto type = origType.getType();
+    if (auto opaqueType = dyn_cast<OpaqueTypeArchetypeType>(type)) {
+      auto conforms = opaqueType->getConformsTo();
+      if (conforms.size() == 1) {
+        auto proto = conforms.front();
+        auto equatable = tc.Context.getProtocol(KnownProtocolKind::Equatable);
+        if (proto == equatable && substType == tc.Context.TheEmptyTupleType)
+          isOpaqueEquatableVoid = true;
+      }
+    }
+  }
```

Comment:
This should just be added to `matchesTuple`; the `::Type` cases should allow `OpaqueTypeArchetypeType`s.  This hasn't been an issue before because tuples haven't had any protocol conformances.